README.md
# Getting help

Need help? Not to worry, we have you covered.

We have a couple resources designed to help you learn, we suggest starting with the tutorial and from there, moving on to the Pony Patterns book. Additionally, standard library documentation is available online.

* [Tutorial](http://tutorial.ponylang.org).
* [Pony Patterns](http://patterns.ponylang.org) cookbook is in progress
* [Standard library docs](http://ponylang.github.io/ponyc/).

If you are looking for an answer "right now", we suggest you give our IRC channel a try. It's #ponylang on Freenode. If you ask a question, be sure to hang around until you get an answer. If you don't get one, or IRC isn't your thing, we have a friendly mailing list you can try. Whatever your question is, it isn't dumb, and we won't get annoyed.

* [Mailing list](mailto:pony+user@groups.io).
* [IRC](https://webchat.freenode.net/?channels=%23ponylang).

Think you've found a bug? Check your understanding first by writing the mailing list. Once you know it's a bug, open an issue.

* [Open an issue](https://github.com/ponylang/ponyc/issues)

# Editor support

* Sublime Text: [Pony Language](https://packagecontrol.io/packages/Pony%20Language)
* Atom: [language-pony](https://atom.io/packages/language-pony)
* Visual Studio: [VS-pony](https://github.com/ponylang/VS-pony)
* Visual Studio Code: [vscode-pony](https://marketplace.visualstudio.com/items?itemName=npruehs.pony)
* Vim:
    - [vim-pony](https://github.com/jakwings/vim-pony)
    - [pony.vim](https://github.com/dleonard0/pony-vim-syntax)
* Emacs:
    - [ponylang-mode](https://github.com/seantallen/ponylang-mode)
    - [flycheck-pony](https://github.com/rmloveland/flycheck-pony)
    - [pony-snippets](https://github.com/SeanTAllen/pony-snippets)
* BBEdit: [bbedit-pony](https://github.com/TheMue/bbedit-pony)

# Installation

## Using Docker

Want to use the latest revision of Pony source, but don't want to build from source yourself? You can run the `ponylang/ponyc` Docker container, which is created from an automated build at each commit to master.

You'll need to install Docker using [the instructions here](https://docs.docker.com/engine/installation/). Then you can pull the latest `ponylang/ponyc` image using this command:

```bash
$ docker pull ponylang/ponyc:latest
```

Then you'll be able to run `ponyc` to compile a Pony program in a given directory, running a command like this:

```bash
$ docker run -v /path/to/my-code:/src/main ponylang/ponyc
```

If you're unfamiliar with Docker, remember to ensure that whatever path you provide for `/path/to/my-code` is a full path name and not a relative path, and also note the lack of a closing slash, `/`, at the *end* of the path name.

Note that if your host doesn't match the docker container, you'll probably have to run the resulting program inside the docker container as well, using a command like this:

```bash
$ docker run -v /path/to/my-code:/src/main ponylang/ponyc ./main
```

If you're using `docker-machine` instead of native docker, make sure you aren't using [an incompatible version of Virtualbox](#virtualbox).


## Linux using an RPM package (via Bintray)

For Red Hat, CentOS, Oracle Linux, or Fedora Linux, the `master` and `release` branches are packaged and available on Bintray ([pony-language/ponyc-rpm](https://bintray.com/pony-language/ponyc-rpm)).

To install release builds via Yum:
```bash
wget https://bintray.com/pony-language/ponyc-rpm/rpm -O bintray-pony-language-ponyc-rpm.repo
sudo mv bintray-pony-language-ponyc-rpm.repo /etc/yum.repos.d/

sudo yum install ponyc-release
```

Or, for master builds:
```bash
yum install ponyc-master
```

## Linux using a DEB package (via Bintray)

For Ubuntu or Debian Linux, the `master` and `release` branches are packaged and available on Bintray ([pony-language/ponyc-debian](https://bintray.com/pony-language/ponyc-debian)).

To install release builds via Apt:

```bash
echo "deb https://dl.bintray.com/pony-language/ponyc-debian pony-language main" | sudo tee -a /etc/apt/sources.list
sudo apt-get update
sudo apt-get install ponyc-release
```


Or, for master builds:
```
sudo apt-get install ponyc-master
```

## Windows using ZIP (via Bintray)

For Windows, the `master` and `release` branches are packaged and available on Bintray ([pony-language/ponyc-win](https://bintray.com/pony-language/ponyc-win)):

```powershell
Invoke-WebRequest -Uri https://dl.bintray.com/pony-language/ponyc-win/ponyc-VERSION.zip -UseBasicParsing -OutFile ponyc-VERSION.zip
7z x .\ponyc-VERSION.zip
.\ponyc-VERSION\ponyc\bin\ponyc.exe --version
```

Windows 10 users will need to install the Windows 10 SDK in order to build programs with ponyc. It can be downloaded [from Microsoft](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk).

## Mac OS X using [Homebrew](http://brew.sh)

```bash
$ brew update
$ brew install ponyc
```

## Arch Linux

```bash
pacman -S ponyc
```

## Gentoo Linux

```bash
$ layman -a stefantalpalaru
$ emerge dev-lang/pony
```

A live ebuild is also available in the [overlay](https://github.com/stefantalpalaru/gentoo-overlay) (dev-lang/pony-9999) and for Vim users there's app-vim/pony-syntax.

# Building ponyc from source

First of all, you need a compiler with decent C11 support. The following compilers are supported, though we recommend to use the most recent versions.

- GCC >= 4.7
- Clang >= 3.3
- MSVC >= 2013
- XCode Clang >= 6.0

Pony requires one of the following versions of LLVM:

- 3.6.2
- 3.7.1
- 3.8.1
- 3.9.0

Compiling Pony is only possible on x86 and ARM (either 32 or 64 bits).

## Building on Linux

Get Pony-Sources from Github (More Information about Set Up Git https://help.github.com/articles/set-up-git/ ):
```bash
$ sudo apt install git
$ git clone git://github.com/ponylang/ponyc
```

[![Linux and OS X](https://travis-ci.org/ponylang/ponyc.svg?branch=master)](https://travis-ci.org/ponylang/ponyc)

### Arch

```
pacman -S llvm make ncurses openssl pcre2 zlib
```

To build ponyc and compile helloworld:

```bash
$ make
$ ./build/release/ponyc examples/helloworld
```

If you get errors like

```bash
/usr/bin/ld.gold: error: ./fb.o: requires dynamic R_X86_64_32 reloc against 'Array_String_val_Trace' which may overflow at runtime; recompile with -fPIC
```

try running `ponyc` with the `--pic` flag.

```bash
$ ./build/release/ponyc --pic examples/helloworld
```

### Debian Jessie

Add the following to `/etc/apt/sources`:

```
deb http://llvm.org/apt/jessie/ llvm-toolchain-jessie-3.8 main
deb-src http://llvm.org/apt/jessie/ llvm-toolchain-jessie-3.8 main
```

Install the LLVM toolchain public GPG key, update `apt` and install packages:

```bash
$ wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key|sudo apt-key add -
$ sudo apt-get update
$ sudo apt-get install make gcc g++ git zlib1g-dev libncurses5-dev \
                       libssl-dev llvm-3.8-dev
```

Debian Jessie and some other Linux distributions don't include pcre2 in their package manager. pcre2 is used by the Pony regex package. To download and build pcre2 from source:

```bash
$ wget ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre2-10.21.tar.bz2
$ tar xvf pcre2-10.21.tar.bz2
$ cd pcre2-10.21
$ ./configure --prefix=/usr
$ make
$ sudo make install
```

To build ponyc, compile and run helloworld:

```bash
$ cd ~/ponyc/
$ make
$ ./build/release/ponyc examples/helloworld
$ ./helloworld
```

### Ubuntu (14.04, 15.10, 16.04)

You should install prebuilt Clang 3.8 from the [LLVM download page](http://llvm.org/releases/download.html#3.8.0) under Pre-Built Binaries:

```bash
$ sudo apt-get update
$ sudo apt-get install -y build-essential git zlib1g-dev libncurses5-dev libssl-dev llvm-dev
```

Ubuntu and some other Linux distributions don't include pcre2 in their package manager. pcre2 is used by the Pony regex package. To download and build pcre2 from source:

```bash
$ wget ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre2-10.21.tar.bz2
$ tar xvf pcre2-10.21.tar.bz2
$ cd pcre2-10.21
$ ./configure --prefix=/usr
$ make
$ sudo make install
```

To build ponyc, compile and run helloworld:

```bash
$ cd ~/ponyc/
$ make
$ ./build/release/ponyc examples/helloworld
$ ./helloworld
```

### Other Linux distributions

You need to have the development versions of the following installed:

* LLVM 3.6.2, 3.7.1, 3.8.1.
* zlib
* ncurses
* pcre2
* libssl

If your distribution doesn't have a package for prce2, you will need to download and build it from source:

```bash
$ wget ftp://ftp.csx.cam.ac.uk/pub/software/programming/pcre/pcre2-10.21.tar.bz2
$ tar xvf pcre2-10.21.tar.bz2
$ cd pcre2-10.21
$ ./configure --prefix=/usr
$ make
$ sudo make install
```

Finally to build ponyc, compile and run the hello world app:

```bash
$ cd ~/ponyc/
$ make
$ ./build/release/ponyc examples/helloworld
$ ./helloworld
```

## Building on FreeBSD

First, install the required dependencies:

```bash
$ sudo pkg install gmake
$ sudo pkg install llvm38
$ sudo pkg install pcre2
$ sudo pkg install libunwind
```

This will build ponyc and compile helloworld:

```bash
$ gmake
$ ./build/release/ponyc examples/helloworld
```

Please note that on 32-bit X86, using LLVM 3.7.1 or 3.8.1 on FreeBSD currently produces executables that don't run. Please use LLVM 3.6.2. 64-bit X86 does not have this problem, and works fine with LLVM 3.7.1 and 3.8.1.

## Building on Mac OS X
[![Linux and OS X](https://travis-ci.org/ponylang/ponyc.svg?branch=master)](https://travis-ci.org/ponylang/ponyc)

You'll need llvm 3.6.2, 3.7.1, or 3.8.1 and the pcre2 library to build Pony. You can use either homebrew or MacPorts to install these dependencies.

Installation via [homebrew](http://brew.sh):
```
$ brew update
$ brew install homebrew/versions/llvm38 pcre2 libressl
```

Installation via [MacPorts](https://www.macports.org):
```
$ sudo port install llvm-3.8 pcre2 libressl
$ sudo port select --set llvm mp-llvm-3.8
```

Launch the build with `make` after installing the dependencies:
```
$ make
$ ./build/release/ponyc examples/helloworld
```

## Building on Windows
[![Windows](https://ci.appveyor.com/api/projects/status/kckam0f1a1o0ly2j?svg=true)](https://ci.appveyor.com/project/sylvanc/ponyc)

_Note: it may also be possible (as tested on build 14372.0 of Windows 10) to build Pony using the [Ubuntu 14.04](#ubuntu-1404-1510-1604) instructions inside [Bash on Ubuntu on Windows](https://msdn.microsoft.com/en-us/commandline/wsl/about)._

The LLVM prebuilt binaries for Windows do NOT include the LLVM development tools and libraries. Instead, you will have to build and install LLVM 3.7.1 or 3.8.1 from source. You will need to make sure that the path to LLVM/bin (location of llvm-config) is in your PATH variable.

LLVM recommends using the GnuWin32 unix tools; your mileage may vary using MSYS or Cygwin.

- Install GnuWin32 using the [GetGnuWin32](http://getgnuwin32.sourceforge.net/) tool.
- Install [Python](https://www.python.org/downloads/release/python-351/) (3.5 or 2.7).
- Install [CMake](https://cmake.org/download/).
- Get the LLVM source (e.g. 3.7.1 is at [3.7.1](http://llvm.org/releases/3.7.1/llvm-3.7.1.src.tar.xz)).
- Make sure you have VS2015 with the C++ tools installed.
- Generate LLVM VS2015 configuration with CMake. You can use the GUI to configure and generate the VS projects; make sure you use the 64-bit generator (**Visual Studio 14 2015 Win64**), and set the `CMAKE_INSTALL_PREFIX` to where you want LLVM to live.
- Open the LLVM.sln in Visual Studio 2015 and build the INSTALL project in the LLVM solution in Release mode.

Building Pony requires [Premake 5](https://premake.github.io).

- Get the [PreMake 5](https://premake.github.io/download.html#v5) executable.
- Get the [PonyC source](https://github.com/ponylang/ponyc).
- Run `premake5.exe --with-tests --to=..\vs vs2015` to generate the PonyC
  solution.
- Build ponyc.sln in Release mode.

In order to run the pony compiler, you'll need a few libraries in your environment (pcre2, libssl, libcrypto).

There is a third-party utility that will get the libraries and set up your environment:

- Install [7-Zip](http://www.7-zip.org/a/7z1514-x64.exe), make sure it's in your PATH.
- Open a **VS2015 x64 Native Tools Command Prompt** (things will not work correctly otherwise!) and run:

```bash
$ git clone git@github.com:kulibali/ponyc-windows-libs.git
$ cd ponyc-windows-libs
$ .\getlibs.bat
$ .\setenv.bat
```

Now you can run the pony compiler and tests:

```bash
$ cd path_to_pony_source
$ build\release\testc.exe
$ build\release\testrt.exe
$ build\release\ponyc.exe -d -s packages\stdlib
$ .\stdlib
```

## Building with link-time optimisation (LTO)

You can enable LTO when building the compiler in release mode. There are slight differences between platforms so you'll need to do a manual setup. LTO is enabled by setting `lto` to `yes` in the build command line:

```bash
$ make lto=yes
```

If the build fails, you have to specify the LTO plugin for your compiler in the `LTO_PLUGIN` variable. For example:

```bash
$ make LTO_PLUGIN=/usr/lib/LLVMgold.so
```

Refer to your compiler documentation for the plugin to use in your case.

## VirtualBox

Pony binaries can trigger illegal instruction errors under VirtualBox 4.x, for at least the x86_64 platform and possibly others.

Use VirtualBox 5.x to avoid possible problems.

## AVX2 Support

The Pony prebuilt binaries trigger illegal instruction errors under CPUs without [AVX2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#Advanced_Vector_Extensions_2) support.

## Building Pony on Non-x86 platforms

On ARM platforms, the default gcc architecture specification used in the Makefile of _native_ does not work correctly, and can even result in the gcc compiler crashing.  You will have to override the compiler architecture specification on the _make_ command line.  For example, on a RaspberryPi2 you would say:
```bash
$ make arch=armv7
```

To get a complete list of acceptable architecture names, use the gcc command:

```bash
gcc -march=none
```

This will result in an error message plus a listing off all architecture types acceptable on your platform.


CHANGELOG.md
# Change Log

All notable changes to the Pony compiler and standard library will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/) and [Keep a CHANGELOG](http://keepachangelog.com/).

## [unreleased] - unreleased

### Fixed

- Set null terminator in String.from_iso_array (issue #1435) (PR #1436)
### Added

- Extra DTrace/SystemTap probes concerning scheduling.

### Changed

- Update from_array to prevent a copy (issue #1097) (PR #1423)

## [0.9.0] - 2016-11-11

### Fixed

- Stop leaking memory during serialization (issue #1413) (PR #1414)
- Fixed compiler segmentation fault when given an invalid target triple. (PR #1406)
- Fixed error message when no type arguments are given (issue #1396) (PR #1397)
- Fixed compiler assert failure when constructor is called on type intersection (issue #1398) (PR #1401)
- Fix compiler assert fail on circular type inference error (issue #1334) (PR #1339)
- Performance problem in the scheduler queue when running with many threads (issue #1404)
- Invalid name mangling in generated C headers (issue #1377)

### Added

- Method chaining (RFC #25) (PR #1411)
- Iter class methods `all`, `any`, `collect`, `count`, `find`, `last`, `nth`, `run`, `skip`, `skip_while`, `take`, `take_while` (issue #1370)
- Output of `ponyc --version` shows C compiler used to build pony (issue #1245)
- Makefile detects `llvmconfig39` in addition to `llvm-config-3.9` (#1379)
- LLVM 3.9 support

### Changed

- Changed lambda literal syntax to be more concise (issue #1391) (PR #1400)

## [0.8.0] - 2016-10-27

### Fixed

- Link the correct version of `libponyrt` when compiling with `--pic` on Linux (issue #1359)

### Added

- Runtime function `pony_send_next`. This function can help optimise some message sending scenarios.
- Floating point `min_normalised`. The function returns the smallest normalised positive number, as `min_value` used to do (issue #1351)

### Changed

- Floating point `min_value` now returns the smallest negative number instead of the smallest normalised positive number (issue #1351)

## [0.7.0] - 2016-10-22

### Fixed

- Concatenate docstrings from case methods (issue #575).

### Added

- TCP read and write backpressure hooks in `TCPConnection` (issue #1311)
- Allow TCP notifiers to cause connections to yield while receiving (issue #1343)

### Changed

- `break` without a value now generates its value from the `else` branch of a loop instead of being an implicit `break None`.
- The `for` loop will now break out of the loop instead of continuing with the following iterations if `Iterator.next` errors.

## [0.6.0] - 2016-10-20

### Fixed

- Compiling ponyrt with Clang versions >= 3.3, < 3.6.
- Restrict mutable tuple recovery to maintain reference capability security (issue #1123)
- Crash in the runtime scheduler queues

### Added

- DTrace and SystemTap support - `use=dtrace`

### Changed

- Replaces `use=telemetry` by DTrace/SystemTap scripts
- `String.cstring()` now always returns a null-terminated string
  (which may result in a copy) while `cpointer()` (also available on
  `Array` objects) returns a pointer to the underlying array as-is
  (issue #1309).

## [0.5.1] - 2016-10-15

### Fixed

- `SSLConnection` ignoring the `sent` notifier method (issue #1268)
- Runtime crash in the runtime scheduler queues (issue #1319)

## [0.5.0] - 2016-10-09

### Fixed

- Memory copy bounds for `String.clone` (issue #1289).
- Security issues in `ProcessMonitor` (issue #1180)
- `SSLConnection` bugs due to missing `sentv` notify method (issue #1282)

### Added

- `Iter` class (issue #1267)
- read_until method on buffered.Reader (RFC 0013)
- `format` package (issue #1285)

### Changed

- `Stringable` interface no longer involves formatting (issue #1285)
- Remove unused error types from ProcessError (issue #1293)
- HTML documentation for expanded union types now adds line breaks to improve readability (issue #1263)

## [0.4.0] - 2016-09-26

### Fixed

- Unexpected message ordering in `ProcessManager` (issue #1265)

### Added

- TCP writev performance improvement by avoiding throwing errors

## [0.3.3] - 2016-09-23

### Fixed

- Incorrect build number generated on Windows when building from non-git directory.
- Stop generating `llvm.invariant.load` for fields of `val` references.
- Embedded fields construction through tuples.

### Added
- Improved error handling for `files` package.

- Improved error handling for `files` package.
- ProcessMonitor.expect
- ProcessNotify.created
- ProcessNotify.expect

### Changed

- On Linux and FreeBSD, ponyc now uses $CC as the linker if the environment variable is defined.

## [0.3.2] - 2016-09-18

### Fixed

- The `ponyc` version is now consistently set from the VERSION file.
- Stop generating `llvm.invariant.load` intrinsic for "let" references, as these don't necessarily match the semantics of that intrinsic.

### Changed

- The `setversion` and `release` commands have been removed from `Makefile`.
- LTO is again enabled by default on OSX
- make now builds a `release` rather than `debug` build by default

## [0.3.1] - 2016-09-14

### Fixed

- Make sure all scheduler threads are pinned to CPU cores; on Linux/FreeBSD this wasn't the case for the main thread.
- Account for both hyperthreading and NUMA locality when assigning scheduler threads to cores on Linux.
- Stop generating `llvm.invariant.start` intrinsic. It was causing various problems in code generation.
- Buffer overflow triggerable by very long `ponyc` filename (issue #1177).
- Assertion failure in optimisation passes.
- Race condition in scheduler queues on weakly-ordered architectures.
- Issue #1212 by reverting commit e56075d46d7d9e1d8c5e8be7ed0506ad2de98734

### Added

- `--ponypinasio` runtime option for pinning asio thread to a cpu core.
- `--ponynopin` runtime option for not pinning any threads at all.

### Changed

- Path.base now provides option to omit the file extension from the result.

src/ponyc/main.c
#include "../libponyc/ponyc.h"
#include "../libponyc/ast/parserapi.h"
#include "../libponyc/ast/bnfprint.h"
#include "../libponyc/pkg/package.h"
#include "../libponyc/pkg/buildflagset.h"
#include "../libponyc/pass/pass.h"
#include "../libponyc/ast/stringtab.h"
#include "../libponyc/ast/treecheck.h"
#include <platform.h>
#include "../libponyrt/options/options.h"
#include "../libponyrt/mem/pool.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#ifdef PLATFORM_IS_POSIX_BASED
#  include <sys/ioctl.h>
#  include <unistd.h>
#endif

enum
{
  OPT_VERSION,
  OPT_DEBUG,
  OPT_BUILDFLAG,
  OPT_STRIP,
  OPT_PATHS,
  OPT_OUTPUT,
  OPT_LIBRARY,
  OPT_EXPORT,
  OPT_RUNTIMEBC,
  OPT_PIC,
  OPT_DOCS,

  OPT_SAFE,
  OPT_IEEEMATH,
  OPT_CPU,
  OPT_FEATURES,
  OPT_TRIPLE,
  OPT_STATS,
  OPT_LINK_ARCH,
  OPT_LINKER,

  OPT_VERBOSE,
  OPT_PASSES,
  OPT_AST,
  OPT_ASTPACKAGE,
  OPT_TRACE,
  OPT_WIDTH,
  OPT_IMMERR,
  OPT_VERIFY,
  OPT_FILENAMES,
  OPT_CHECKTREE,

  OPT_BNF,
  OPT_ANTLR,
  OPT_ANTLRRAW,
  OPT_EXTFUN
};

static opt_arg_t args[] =
{
  {"version", 'v', OPT_ARG_NONE, OPT_VERSION},
  {"debug", 'd', OPT_ARG_NONE, OPT_DEBUG},
  {"define", 'D', OPT_ARG_REQUIRED, OPT_BUILDFLAG},
  {"strip", 's', OPT_ARG_NONE, OPT_STRIP},
  {"path", 'p', OPT_ARG_REQUIRED, OPT_PATHS},
  {"output", 'o', OPT_ARG_REQUIRED, OPT_OUTPUT},
  {"library", 'l', OPT_ARG_NONE, OPT_LIBRARY},
  {"export", 'e', OPT_ARG_NONE, OPT_EXPORT},
  {"runtimebc", '\0', OPT_ARG_NONE, OPT_RUNTIMEBC},
  {"pic", '\0', OPT_ARG_NONE, OPT_PIC},
  {"docs", 'g', OPT_ARG_NONE, OPT_DOCS},

  {"safe", '\0', OPT_ARG_OPTIONAL, OPT_SAFE},
  {"ieee-math", '\0', OPT_ARG_NONE, OPT_IEEEMATH},
  {"cpu", '\0', OPT_ARG_REQUIRED, OPT_CPU},
  {"features", '\0', OPT_ARG_REQUIRED, OPT_FEATURES},
  {"triple", '\0', OPT_ARG_REQUIRED, OPT_TRIPLE},
  {"stats", '\0', OPT_ARG_NONE, OPT_STATS},
  {"link-arch", '\0', OPT_ARG_REQUIRED, OPT_LINK_ARCH},
  {"linker", '\0', OPT_ARG_REQUIRED, OPT_LINKER},

  {"verbose", 'V', OPT_ARG_REQUIRED, OPT_VERBOSE},
  {"pass", 'r', OPT_ARG_REQUIRED, OPT_PASSES},
  {"ast", 'a', OPT_ARG_NONE, OPT_AST},
  {"astpackage", '\0', OPT_ARG_NONE, OPT_ASTPACKAGE},
  {"trace", 't', OPT_ARG_NONE, OPT_TRACE},
  {"width", 'w', OPT_ARG_REQUIRED, OPT_WIDTH},
  {"immerr", '\0', OPT_ARG_NONE, OPT_IMMERR},
  {"verify", '\0', OPT_ARG_NONE, OPT_VERIFY},
  {"files", '\0', OPT_ARG_NONE, OPT_FILENAMES},
  {"checktree", '\0', OPT_ARG_NONE, OPT_CHECKTREE},

  {"bnf", '\0', OPT_ARG_NONE, OPT_BNF},
  {"antlr", '\0', OPT_ARG_NONE, OPT_ANTLR},
  {"antlrraw", '\0', OPT_ARG_NONE, OPT_ANTLRRAW},
  {"extfun", '\0', OPT_ARG_NONE, OPT_EXTFUN},
  OPT_ARGS_FINISH
};

static void usage()
{
  printf(
    "ponyc [OPTIONS] <package directory>\n"
    "\n"
    "The package directory defaults to the current directory.\n"
    "\n"
    "Options:\n"
    "  --version, -v   Print the version of the compiler and exit.\n"
    "  --debug, -d     Don't optimise the output.\n"
    "  --define, -D    Define the specified build flag.\n"
    "    =name\n"
    "  --strip, -s     Strip debug info.\n"
    "  --path, -p      Add an additional search path.\n"
    "    =path         Used to find packages and libraries.\n"
    "  --output, -o    Write output to this directory.\n"
    "    =path         Defaults to the current directory.\n"
    "  --library, -l   Generate a C-API compatible static library.\n"
    "  --export, -e    Generate exported functions.\n"
    "  --runtimebc     Compile with the LLVM bitcode file for the runtime.\n"
    "  --pic           Compile using position independent code.\n"
    "  --docs, -g      Generate code documentation.\n"
    "\n"
    "Rarely needed options:\n"
    "  --safe          Allow only the listed packages to use C FFI.\n"
    "    =package      With no packages listed, only builtin is allowed.\n"
    "  --ieee-math     Force strict IEEE 754 compliance.\n"
    "  --cpu           Set the target CPU.\n"
    "    =name         Default is the host CPU.\n"
    "  --features      CPU features to enable or disable.\n"
    "    =+this,-that  Use + to enable, - to disable.\n"
    "  --triple        Set the target triple.\n"
    "    =name         Defaults to the host triple.\n"
    "  --stats         Print some compiler stats.\n"
    "  --link-arch     Set the linking architecture.\n"
    "    =name         Default is the host architecture.\n"
    "  --linker        Set the linker command to use.\n"
    "    =name         Default is the compiler used to compile ponyc.\n"
    "\n"
    "Debugging options:\n"
    "  --verbose, -V   Verbosity level.\n"
    "    =0            Only print errors.\n"
    "    =1            Print info on compiler stages.\n"
    "    =2            More detailed compilation information.\n"
    "    =3            External tool command lines.\n"
    "    =4            Very low-level detail.\n"
    "  --pass, -r      Restrict phases.\n"
    "    =parse\n"
    "    =syntax\n"
    "    =sugar\n"
    "    =scope\n"
    "    =import\n"
    "    =name\n"
    "    =flatten\n"
    "    =traits\n"
    "    =docs\n"
    "    =expr\n"
    "    =verify\n"
    "    =final\n"
    "    =reach\n"
    "    =paint\n"
    "    =ir           Output LLVM IR.\n"
    "    =bitcode      Output LLVM bitcode.\n"
    "    =asm          Output assembly.\n"
    "    =obj          Output an object file.\n"
    "    =all          The default: generate an executable.\n"
    "  --ast, -a       Output an abstract syntax tree for the whole program.\n"
    "  --astpackage    Output an abstract syntax tree for the main package.\n"
    "  --trace, -t     Enable parse trace.\n"
    "  --width, -w     Width to target when printing the AST.\n"
    "    =columns      Defaults to the terminal width.\n"
    "  --immerr        Report errors immediately rather than deferring.\n"
    "  --verify        Verify LLVM IR.\n"
    "  --extfun        Set function default linkage to external.\n"
    "  --files         Print source file names as each is processed.\n"
    "  --bnf           Print out the Pony grammar as human readable BNF.\n"
    "  --antlr         Print out the Pony grammar as an ANTLR file.\n"
    "\n"
    "Runtime options for Pony programs (not for use with ponyc):\n"
    "  --ponythreads   Use N scheduler threads. Defaults to the number of\n"
    "                  cores (not hyperthreads) available.\n"
    "  --ponycdmin     Defer cycle detection until 2^N actors have blocked.\n"
    "                  Defaults to 2^4.\n"
    "  --ponycdmax     Always cycle detect when 2^N actors have blocked.\n"
    "                  Defaults to 2^18.\n"
    "  --ponycdconf    Send cycle detection CNF messages in groups of 2^N.\n"
    "                  Defaults to 2^6.\n"
    "  --ponygcinitial Defer garbage collection until an actor is using at\n"
    "                  least 2^N bytes. Defaults to 2^14.\n"
    "  --ponygcfactor  After GC, an actor will next be GC'd at a heap memory\n"
    "                  usage N times its current value. This is a floating\n"
    "                  point value. Defaults to 2.0.\n"
    "  --ponynoyield   Do not yield the CPU when no work is available.\n"
    "  --ponynoblock   Do not send block messages to the cycle detector.\n"
    "  --ponynopin     Do not pin scheduler threads or the ASIO thread, even\n"
    "                  if --ponypinasio is set.\n"
    "                  threads are pinned to CPUs.\n"
    "  --ponypinasio   Pin the ASIO thread to a CPU the way scheduler\n"
    "                  threads are pinned to CPUs.\n"
    );
}

static size_t get_width()
{
  size_t width = 80;
#ifdef PLATFORM_IS_WINDOWS
  if(_isatty(_fileno(stdout)))
  {
    CONSOLE_SCREEN_BUFFER_INFO info;

    if(GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &info))
    {
      int cols = info.srWindow.Right - info.srWindow.Left + 1;

      if(cols > width)
        width = cols;
    }
  }
#else
  if(isatty(STDOUT_FILENO))
  {
    struct winsize ws;

    if(ioctl(STDOUT_FILENO, TIOCGWINSZ, &ws) == 0)
    {
      if(ws.ws_col > width)
        width = ws.ws_col;
    }
  }
#endif
  return width;
}

static bool compile_package(const char* path, pass_opt_t* opt,
  bool print_program_ast, bool print_package_ast)
{
  ast_t* program = program_load(path, opt);

  if(program == NULL)
    return false;

  if(print_program_ast)
    ast_fprint(stderr, program);

  if(print_package_ast)
    ast_fprint(stderr, ast_child(program));

  bool ok = generate_passes(program, opt);
  ast_free(program);
  return ok;
}

int main(int argc, char* argv[])
{
  stringtab_init();

  pass_opt_t opt;
  pass_opt_init(&opt);

  opt.release = true;
  opt.output = ".";

  ast_setwidth(get_width());
  bool print_program_ast = false;
  bool print_package_ast = false;

  opt_state_t s;
  ponyint_opt_init(args, &s, &argc, argv);

  bool ok = true;
  bool print_usage = false;
  int id;

  while((id = ponyint_opt_next(&s)) != -1)
  {
    switch(id)
    {
      case OPT_VERSION:
        printf("%s [%s]\ncompiled with: llvm %s -- %s\n", PONY_VERSION, PONY_BUILD_CONFIG,
          LLVM_VERSION, BUILD_COMPILER);
        return 0;

      case OPT_DEBUG: opt.release = false; break;
      case OPT_BUILDFLAG: define_build_flag(s.arg_val); break;
      case OPT_STRIP: opt.strip_debug = true; break;
      case OPT_PATHS: package_add_paths(s.arg_val, &opt); break;
      case OPT_OUTPUT: opt.output = s.arg_val; break;
      case OPT_LIBRARY: opt.library = true; break;
      case OPT_EXPORT: opt.export_methods = true; break;
      case OPT_RUNTIMEBC: opt.runtimebc = true; break;
      case OPT_PIC: opt.pic = true; break;
      case OPT_DOCS: opt.docs = true; break;

      case OPT_SAFE:
        if(!package_add_safe(s.arg_val, &opt))
          ok = false;
        break;

      case OPT_IEEEMATH: opt.ieee_math = true; break;
      case OPT_CPU: opt.cpu = s.arg_val; break;
      case OPT_FEATURES: opt.features = s.arg_val; break;
      case OPT_TRIPLE: opt.triple = s.arg_val; break;
      case OPT_STATS: opt.print_stats = true; break;
      case OPT_LINK_ARCH: opt.link_arch = s.arg_val; break;
      case OPT_LINKER: opt.linker = s.arg_val; break;

      case OPT_AST: print_program_ast = true; break;
      case OPT_ASTPACKAGE: print_package_ast = true; break;
      case OPT_TRACE: parse_trace(true); break;
      case OPT_WIDTH: ast_setwidth(atoi(s.arg_val)); break;
      case OPT_IMMERR: errors_set_immediate(opt.check.errors, true); break;
      case OPT_VERIFY: opt.verify = true; break;
      case OPT_EXTFUN: opt.extfun = true; break;
      case OPT_FILENAMES: opt.print_filenames = true; break;
      case OPT_CHECKTREE: opt.check_tree = true; break;

      case OPT_BNF: print_grammar(false, true); return 0;
      case OPT_ANTLR: print_grammar(true, true); return 0;
      case OPT_ANTLRRAW: print_grammar(true, false); return 0;

      case OPT_VERBOSE:
        {
          int v = atoi(s.arg_val);
          if (v >= 0 && v <= 4) {
            opt.verbosity = (verbosity_level)v;
          } else {
            ok = false;
          }
        }
        break;

      case OPT_PASSES:
        if(!limit_passes(&opt, s.arg_val))
        {
          ok = false;
          print_usage = true;
        }
        break;

      default: usage(); return -1;
    }
  }

  for(int i = 1; i < argc; i++)
  {
    if(argv[i][0] == '-')
    {
      printf("Unrecognised option: %s\n", argv[i]);
      ok = false;
      print_usage = true;
    }
  }

#if defined(PLATFORM_IS_WINDOWS)
  opt.strip_debug = true;
#endif

  if(!ok)
  {
    errors_print(opt.check.errors);

    if(print_usage)
      usage();

    return -1;
  }

  if(ponyc_init(&opt))
  {
    if(argc == 1)
    {
      ok &= compile_package(".", &opt, print_program_ast, print_package_ast);
    } else {
      for(int i = 1; i < argc; i++)
        ok &= compile_package(argv[i], &opt, print_program_ast,
          print_package_ast);
    }
  }

  if(!ok && errors_get_count(opt.check.errors) == 0)
    printf("Error: internal failure not reported\n");

  ponyc_shutdown(&opt);
  pass_opt_done(&opt);

  return ok ? 0 : -1;
}
